src/devices/KeyMatrix/KeyMatrix.cs

```diff
@@ -0,0 +1,179 @@
+        private GpioController _gpioController;
+        private PinValue[] _buttonValues;
+        private bool _pinsOpened;
+        private int _currentOutput = 0;
+
+        /// <summary>
+        /// Initialize key matrix
+        /// </summary>
+        /// <param name="gpioController">GPIO controller</param>
+        /// <param name="outputPins">Output pins</param>
+        /// <param name="inputPins">Input pins</param>
+        /// <param name="scanInterval">Scanning interval in milliseconds</param>
+        public KeyMatrix(GpioController gpioController, IEnumerable<int> outputPins, IEnumerable<int> inputPins, int scanInterval)
+        {
+            _gpioController = gpioController ?? throw new ArgumentNullException(nameof(gpioController));
+
+            if (outputPins == null)
+            {
+                throw new ArgumentNullException(nameof(outputPins));
+            }
+
+            if (!outputPins.Any())
+            {
+                throw new ArgumentOutOfRangeException(nameof(outputPins), "The number of outputs must be at least 1");
+            }
+
+            if (inputPins == null)
+            {
+                throw new ArgumentNullException(nameof(inputPins));
+            }
+
+            if (!inputPins.Any())
+            {
+                throw new ArgumentOutOfRangeException(nameof(inputPins), "The number of inputs must be at least 1");
+            }
+
+            if (scanInterval < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(scanInterval), "Scanning interval must be 0 or positive");
+            }
+
+            _outputPins = outputPins.ToArray();
+            _inputPins = inputPins.ToArray();
+            _buttonValues = new PinValue[_outputPins.Length * _inputPins.Length];
+            _pinsOpened = false;
+
+            ScanInterval = scanInterval;
+
+            OpenPins();
+        }
+
+        /// <summary>
+        /// Blocks execution until a key event is received
+        /// </summary>
+        public KeyMatrixEvent ReadKey()
+        {
+            KeyMatrixEvent args = null;
+            _currentOutput--;
+
+            do
+            {
+                Thread.Sleep(ScanInterval);
+
+                _currentOutput = (_currentOutput + 1) % _outputPins.Length;
+                _gpioController.Write(_outputPins[_currentOutput], PinValue.High);
+
+                for (var i = 0; i < _inputPins.Length; i++)
+                {
+                    int index = _currentOutput * _inputPins.Length + i;
+
+                    PinValue oldValue = _buttonValues[index];
+                    PinValue newValue = _gpioController.Read(_inputPins[i]);
+                    _buttonValues[index] = newValue;
+
+                    if (newValue != oldValue)
+                    {
+                        args = new KeyMatrixEvent(newValue == PinValue.High ? PinEventTypes.Rising : PinEventTypes.Falling, _currentOutput, i);
+                        return args;
```

Comment:
should you also call `_gpioController.Write(_outputPins[_currentOutput], PinValue.Low);` here?